Analysis/Tasks/pidTOF_split.cxx

```diff
@@ -58,47 +73,68 @@ struct pidTOFTaskPerParticle {
     ccdb->setCreatedNotAfter(std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count());
     //
     const std::vector<float> p = {0.008, 0.008, 0.002, 40.0};
-    resp.SetParameters(DetectorResponse::kSigma, p);
+    response.SetParameters(DetectorResponse::kSigma, p);
     const std::string fname = paramfile.value;
     if (!fname.empty()) { // Loading the parametrization from file
-      resp.LoadParamFromFile(fname.data(), sigmaname.value, DetectorResponse::kSigma);
+      response.LoadParamFromFile(fname.data(), sigmaname.value, DetectorResponse::kSigma);
     } else { // Loading it from CCDB
       const std::string path = "Analysis/PID/TOF";
-      resp.LoadParam(DetectorResponse::kSigma, ccdb->getForTimeStamp<Parametrization>(path + "/" + sigmaname.value, timestamp.value));
+      response.LoadParam(DetectorResponse::kSigma, ccdb->getForTimeStamp<Parametrization>(path + "/" + sigmaname.value, timestamp.value));
     }
   }
 
+  template <o2::track::PID::ID pid>
+  using ResponseImplementation = tof::ExpTimes<Coll::iterator, Trks::iterator, pid>;
   void process(Coll const& collisions, Trks const& tracks)
   {
-    constexpr tof::ExpTimes<Coll::iterator, Trks::iterator, pid_type> resp_PID = tof::ExpTimes<Coll::iterator, Trks::iterator, pid_type>();
+    constexpr auto responseEl = ResponseImplementation<PID::Electron>();
+    constexpr auto responseMu = ResponseImplementation<PID::Muon>();
+    constexpr auto responsePi = ResponseImplementation<PID::Pion>();
+    constexpr auto responseKa = ResponseImplementation<PID::Kaon>();
+    constexpr auto responsePr = ResponseImplementation<PID::Proton>();
+    constexpr auto responseDe = ResponseImplementation<PID::Deuteron>();
+    constexpr auto responseTr = ResponseImplementation<PID::Triton>();
+    constexpr auto responseHe = ResponseImplementation<PID::Helium3>();
+    constexpr auto responseAl = ResponseImplementation<PID::Alpha>();
+
+    // Prepare memory for enabled tables
+    auto reserveTable = [&tracks](const Configurable<int>& flag, auto& table) {
+      if (flag.value) {
+        table.reserve(tracks.size());
+      }
+    };
+    reserveTable(pidEl, tablePIDEl);
+    reserveTable(pidMu, tablePIDMu);
+    reserveTable(pidPi, tablePIDPi);
+    reserveTable(pidKa, tablePIDKa);
+    reserveTable(pidPr, tablePIDPr);
+    reserveTable(pidDe, tablePIDDe);
+    reserveTable(pidTr, tablePIDTr);
+    reserveTable(pidHe, tablePIDHe);
+    reserveTable(pidAl, tablePIDAl);
 
-    tofpid.reserve(tracks.size());
-    for (auto const& trk : tracks) {
-      tofpid(resp_PID.GetExpectedSigma(resp, trk.collision(), trk),
-             resp_PID.GetSeparation(resp, trk.collision(), trk));
+    for (auto const& trk : tracks) { // Loop on Tracks
+      // Check and fill enabled tables
+      auto makeTable = [&trk](const Configurable<int>& flag, auto& table, const DetectorResponse& response, const auto& responsePID) {
+        if (flag.value) {
```

Comment:
Can we put the loop into the lambda? In this way we don't evaluate the flag.value for each track, but only once per collision.

Reply:
Very good point! Thanks!